subprojects/core-api/src/main/java/org/gradle/api/provider/Property.java

```diff
@@ -77,29 +105,47 @@
+     * <p>This method can be used to specify that the property does not have a default value, by passing it
+     * {@code null}.</p>
+     *
+     * @param value The convention value, or {@code null} if the property should have no default value.
      * @return this
      * @since 5.1
      */
     Property<T> convention(@Nullable T value);
 
     /**
-     * Specifies the provider of the value to use as the convention for this property. The convention is used when no value has been set for this property.
+     * Specifies the provider to be used to query the convention (default value) for this property. If a convention
+     * provider has been set and no explicit value or value provider has been specified, then the convention
+     * provider's value will be returned as the value of the property (when queried).
+     *
+     * <p>The property's convention tracks the convention provider. Whenever the convention's actual value is
+     * needed, the convention provider will be queried anew.</p>
+     *
+     * <p>This method can't be used to specify that a property does not have a default value. Passing in a {@code null}
+     * provider will result in an {@code IllegalArgumentException} being thrown.</p>
      *
-     * @param valueProvider The provider of the value.
+     * @param provider The provider of the property's convention value, can't be null.
      * @return this
      * @since 5.1
      */
-    Property<T> convention(Provider<? extends T> valueProvider);
+    Property<T> convention(Provider<? extends T> provider);
 
     /**
-     * Disallows further changes to the value of this property. Calls to methods that change the value of this property, such as {@link #set(Object)} or {@link #set(Provider)} will fail.
+     * Disallows further changes to the value of this property. Calls to methods that change the value of this property,
+     * such as {@link #set(Object)}, {@link #set(Provider)}, {@link #value(Object)} and {@link #value(Provider)} will fail,
+     * by throwing an {@code IllegalStateException}.
      *
-     * <p>When this property has a value provided by a {@link Provider}, the value of the provider is queried when this method is called and the value of this property set to the result. The value of the provider will no longer be tracked.</p>
+     * <p>When this property's value is specified via a {@link Provider}, this method call will trigger the querying
+     * of the provider and the obtained value will be set as the final value of the property. The value of the provider
+     * will not be tracked further.</p>
      *
-     * <p>Note that although the value of the property will not change, the value itself may be a mutable object. Calling this method does not guarantee that the value will become immutable.</p>
+     * <p>Note that although the value of the property will not change, the value itself may be a mutable object.
+     * Calling this method does not guarantee that the value will become immutable.</p>
```

Comment:
Not directly related to this PR, but it would be convenient to make this API fluid, too.

Reply:
I would have expected to see methods like:

```
default T finalValue() {
   finalizeValue();
   return get();
}

boolean isFinalValue();
```

on this interface, rather than a fluent finalizer.

Reply:
I do see the logic of what you guys are saying. Tom's version would be too much of a change in a public API, I guess. I'm inclined to simply make the current method fluent, like Etienne suggested, but I wonder why `set` hasn't been changed like that, but a fluent variant was added... Will dig into it a bit more, before changing.